### ISSUE TITLE
Compile Describe and Tests as Classes/Objects and Methods

### ISSUE THREAD

**Issue by nscarcella** [uqbar-project/wollok-ts#27]
Currently, describes and tests are separate reified nodes on the AST. The current implementation if fine, but there is quite a lot of redundancy, since Describes, not being Modules, still can define methods and variables. The only difference with regular Modules is that they can define Fixtures, which, if I understand correctly, might be soon deprecated in favor of init blocks. Test have a similar redundancy with methods too, which makes some parts of the pipeline unnecessarily reiterative and the whole model more complex.

So...

Talking with @npasserini, he mentioned that in his head he though of describes and tests as just a separate domain specific syntax to define classes and methods. I was a little resilient at first, but after some time I grow fond of the idea. Maybe we could implement Describes and Tests as just Modules and Methods. We could identify Describes by making them extend a Describe superclass, including a Describe mixin or just using annotations (if we eventually include them). Tests might be identified either with annotations or by a name preface injected at parse. We could alternatively just add a new field in Class and Method nodes.

Mind you **this is not a proposal for a syntax change**. The test writing would be exactly the same, is just a matter of internal representation.

I believe this might be a good idea to keep the project simple, but I would like your feedback. I would be specially interested to know if you all share the idea that describes/test can be think like classes/methods, or if you are thinking of taking the abstractions in a very different direction.

**Comment by nscarcella** [comment]
@fdodino @PalumboN

**Comment by fdodino** [comment]
If we are talking just about implementation, using Describe as Module (Object, Class, ObjectLiteral) and Test as Method is ok. Both will have initialize block (instead of constructor/initialize method / fixture block), so they will be real very similar. 

What I don't want is to change the syntax: I believe X-Unit implementation was born considering current language implementation, instead of using concepts like: test, group of tests, fixture (o example), etc.

**Comment by nscarcella** [comment]
Yes @fdodino, this is just about internal representation. Syntax for tests would remain the same, but we could save on some complexities on the internal model.

**Comment by fdodino** [comment]
@PalumboN @ivojawer yo veo que la actual implementación de Describe es

```ts
export class Describe extends Module(Node) {
  get kind(): 'Describe' { return 'Describe' }
  readonly name!: Name
  readonly members!: List<Field | CodeContainer>
  readonly supertypes: List<ParameterizedType> = [new ParameterizedType({ reference: new Reference({ name: OBJECT_MODULE }) })]

  override parent!: Package

  constructor({ members = [], ...payload }: Payload<Describe, 'name'>) {
    super({ members, ...payload })
  }

  get superclass(): Class { return this.supertypes[0].reference.target! as Class }

  get tests(): List<Test> { return this.members.filter(is(Test)) }

}
```

Me parece que podemos cerrar este issue, qué opinan?
